builtin/providers/aws/data_source_aws_caller_identity.go

```diff
@@ -17,24 +18,40 @@ func dataSourceAwsCallerIdentity() *schema.Resource {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
+
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"user_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
 		},
 	}
 }
 
 func dataSourceAwsCallerIdentityRead(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*AWSClient)
+	client := meta.(*AWSClient).stsconn
+
+	res, err := client.GetCallerIdentity(&sts.GetCallerIdentityInput{})
+	if err != nil {
+		return fmt.Errorf("Error getting Caller Identity: %v", err)
+	}
 
 	log.Printf("[DEBUG] Reading Caller Identity.")
 	d.SetId(time.Now().UTC().String())
 
-	if client.accountid == "" {
+	if *res.Account == "" {
 		log.Println("[DEBUG] No Account ID available, failing")
 		return fmt.Errorf("No AWS Account ID is available to the provider. Please ensure that\n" +
 			"skip_requesting_account_id is not set on the AWS provider.")
 	}
 
-	log.Printf("[DEBUG] Setting AWS Account ID to %s.", client.accountid)
-	d.Set("account_id", meta.(*AWSClient).accountid)
-
+	log.Printf("[DEBUG] Setting AWS Account ID to %s.", *res.Account)
+	d.Set("account_id", *res.Account)
+	d.Set("arn", *res.Arn)
+	d.Set("user_id", *res.UserId)
```

Comment:
A pattern I've seen elsewhere is to just pass in the pointer without dereferencing it, since `d.Set` will automatically unwrap it safely, and thus we avoid crashes if for some reason AWS ends up returning `nil` for any of these.